src/NWAClient.ts

```diff
@@ -0,0 +1,236 @@
+        ? { expires_at: this.options.expiresAt.toString() }
+        : {}),
+      ...(this.options.isolated
+        ? { isolated: this.options.isolated.toString() }
+        : {}),
+      ...(this.options.metadata
+        ? { metadata: JSON.stringify(this.options.metadata) }
+        : {}),
+    });
+
+    return `nostr+walletauth${nwaSchemeSuffix ? `+${nwaSchemeSuffix}` : ""}://${this.options.appPubkey}?${searchParams
+      .toString()
+      .replace(/\+/g, "%20")}`;
+  }
+
+  static parseWalletAuthUrl(walletAuthUrl: string): NWAOptions {
+    if (!walletAuthUrl.startsWith("nostr+walletauth")) {
+      throw new Error(
+        "Unexpected scheme. Should be nostr+walletauth:// or nostr+walletauth+specificapp://",
+      );
+    }
+    // makes it possible to parse with URL in the different environments (browser/node/...)
+    // parses with or without "//"
+    const colonIndex = walletAuthUrl.indexOf(":");
+    walletAuthUrl = walletAuthUrl.substring(colonIndex + 1);
+    if (walletAuthUrl.startsWith("//")) {
+      walletAuthUrl = walletAuthUrl.substring(2);
+    }
+    walletAuthUrl = "http://" + walletAuthUrl;
+
+    const url = new URL(walletAuthUrl);
+
+    const appPubkey = url.host;
+    if (appPubkey?.length !== 64) {
+      throw new Error("Incorrect app pubkey found in auth string");
+    }
+
+    const relayUrl = url.searchParams.get("relay");
+    if (!relayUrl) {
+      throw new Error("No relay URL found in auth string");
+    }
+    const requestMethods = url.searchParams
+      .get("request_methods")
+      ?.split(" ") as Nip47Method[] | undefined;
+    if (!requestMethods?.length) {
+      throw new Error("No request methods found in auth string");
+    }
+
+    const notificationTypes = url.searchParams
+      .get("notification_types")
+      ?.split(" ") as Nip47NotificationType[] | undefined;
+
+    const maxAmountString = url.searchParams.get("max_amount");
+    const expiresAtString = url.searchParams.get("expires_at");
+    const metadataString = url.searchParams.get("metadata");
+
+    return {
+      name: url.searchParams.get("name") || undefined,
+      icon: url.searchParams.get("icon") || undefined,
+      returnTo: url.searchParams.get("return_to") || undefined,
+      relayUrl,
+      appPubkey,
+      requestMethods,
+      notificationTypes,
+      budgetRenewal: url.searchParams.get("budget_renewal") as
+        | BudgetRenewalPeriod
+        | undefined,
+      expiresAt: expiresAtString ? parseInt(expiresAtString) : undefined,
+      maxAmount: maxAmountString ? parseInt(maxAmountString) : undefined,
+      isolated: url.searchParams.get("isolated") === "true",
+      metadata: metadataString ? JSON.parse(metadataString) : undefined,
+    };
+  }
+
+  /**
+   * Waits for a new app connection to be created via NWA (https://github.com/nostr-protocol/nips/pull/851)
+   *
+   * @returns a new NWCClient
+   */
+  async subscribe(args: {
+    onSuccess: (nwcClient: NWCClient) => void;
+  }): Promise<{
```

Comment:
Should we have a timeout with a sufficient default here? Currently it keeps on listening, I don't think that is good design

Reply:
hmm, I am not sure about this, we return an unsubscribe function - the developer can unsubscribe if they want.

It's similar to `setInterval` in javascript, isn't it?